### PR TITLE
Phantom types

### PR DESCRIPTION
**Pull request by hannesm**
this PR introduces phantom types for hostname and service names (instead of passing `~hostname:true` to various functions, there's a ``val host : 'a t -> ([ `host ] t, [> `Msg of string ]) result`` now (and similar `service` and `domain`).

collections (map and set) for hostnames and service names are provided as well

also, ~~concat~~ ``append : 'a t -> 'b t -> [ `domain ] t`` is provided.

/cc @dinosaure @cfcs

### CODE REVIEW

**Comment by dinosaure** [comment]
Seems ok for my case :+1:.

**Comment by cfcs** [comment]
I like it too. :tada:

File: tests.ml [hunk 1]
```diff
@@ -1,28 +1,80 @@
...
   Alcotest.(check bool "_xmpp-server._tcp.foo is a service" true
-              (Domain_name.is_service (n_of_s ~hostname:false "_xmpp-server._tcp.foo"))) ;
+              (is_service (n_of_s "_xmpp-server._tcp.foo"))) ;
   Alcotest.(check bool "_xmpp-server._tcp2.foo is no service" false
-              (Domain_name.is_service (n_of_s ~hostname:false "_xmpp-server._tcp2.foo"))) ;
+              (is_service (n_of_s "_xmpp-server._tcp2.foo"))) ;
   Alcotest.(check bool "_xmpp_server._tcp.foo is no service" false
-              (Domain_name.is_service (n_of_s ~hostname:false "_xmpp_server._tcp.foo"))) ;
+              (is_service (n_of_s "_xmpp_server._tcp.foo"))) ;
   Alcotest.(check bool "_xmpp-server-server._tcp.foo is no service" false
```
Comment (cfcs) [inline tests.ml]: Unrelated to code change, but the comment `_xmpp-server` is different from the used actually used value `_xmpp_server` (underscore/dash)

File: domain_name.mli [hunk 2]
```diff
@@ -72,46 +82,59 @@ val is_service : t -> bool
...
+(** [drop_label_exn ~back ~amount t] is either [t], a domain name with [amount]
     (defaults to 1) labels dropped from the beginning (unless [back] is provided
-    and [true], defaults to [false]).  [drop_labels] applied to [foo.com] is
+    and [true], defaults to [false]).  [drop_label] applied to [foo.com] is
     [com].
 
     @raise Invalid_argument if there are not sufficient labels. *)
 
+val append : 'a t -> 'b t -> ([ `domain ] t, [> `Msg of string ]) result
+(** [append pre post] is [pre ^ "." ^ post]. *)
```
[the remark this answers is not in the document]
Comment (hannesm) [reply domain_name.mli]: thanks fixed

File: tests.ml (hunk 1 above)
Comment (hannesm) [reply tests.ml]: thanks, fixed

File: domain_name.mli (hunk 2 above)
[the remark this answers is not in the document]
Comment (hannesm) [reply domain_name.mli]: all extending operations (prepend_label, ...) share this behaviour, imho it is pretty clear from the types.

File: domain_name.ml [hunk 3]
```diff
@@ -167,11 +192,37 @@ let sub ~subdomain ~domain =
...
   let compare = compare_domain compare_sub
 end
 
-type t = s
+type 'a t = 'a s
+
+module Host_map = struct
+  include Map.Make(Host_ordered)
+
+  let find k m = try Some (find k m) with Not_found -> None
```
[the remark this answers is not in the document]
Comment (hannesm) [reply domain_name.ml]: yes

File: domain_name.mli [hunk 4]
```diff
@@ -72,46 +82,59 @@ val is_service : t -> bool
...
+    and [true], defaults to [false]).  [drop_label] applied to [foo.com] is
     [com]. *)
 
-val drop_labels_exn : ?back:bool -> ?amount:int -> t -> t
-(** [drop_labels_exn ~back ~amount t] is either [t], a domain name with [amount]
+val drop_label_exn : ?back:bool -> ?amount:int -> 'a t -> [ `domain ] t
+(** [drop_label_exn ~back ~amount t] is either [t], a domain name with [amount]
     (defaults to 1) labels dropped from the beginning (unless [back] is provided
-    and [true], defaults to [false]).  [drop_labels] applied to [foo.com] is
+    and [true], defaults to [false]).  [drop_label] applied to [foo.com] is
```
[the remark this answers is not in the document]
Comment (hannesm) [reply domain_name.mli]: addressed, thanks

File: domain_name.mli (hunk 2 above)
[the remark this answers is not in the document]
Comment (hannesm) [reply domain_name.mli]: see 7bd8446

**Comment by cfcs** [comment]
LGTM

**Comment by hannesm** [comment]
merged by hand (renamed `` [`domain] `` to `` [`raw] `` (since `` [`domain] Domain_name.t `` looks slightly ugly)

**Pull request closed without merging**